test/jdk/java/foreign/TestMemoryCopy.java

```diff
@@ -16,106 +16,390 @@
  *  2 along with this work; if not, write to the Free Software Foundation,
  *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
  *
- *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
  *  or visit www.oracle.com if you need additional information or have any
  *  questions.
- *
  */
 
 /*
  * @test
  * @run testng TestMemoryCopy
  */
 
-import jdk.incubator.foreign.MemoryLayouts;
-import jdk.incubator.foreign.MemorySegment;
-import jdk.incubator.foreign.ResourceScope;
-import org.testng.annotations.DataProvider;
-import org.testng.annotations.Test;
+import static org.testng.Assert.assertEquals;
 
 import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
-import java.util.function.IntFunction;
 
-import static org.testng.Assert.*;
+import jdk.incubator.foreign.MemoryCopy;
+import jdk.incubator.foreign.MemoryLayout;
+import jdk.incubator.foreign.MemoryLayouts;
+import jdk.incubator.foreign.MemorySegment;
 
+import jdk.incubator.foreign.ValueLayout;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * These tests exercise the MemoryCopy copyFromArray(...) and copyToArray(...).
+ * To make these tests more challenging the segment is a view of the given array,
+ * which makes the copy operations overlapping self-copies.  Thus, this checks the claim:
```

Comment:
I'm not sure I understand this claim; it seems to me that the test sets up an array view segment, but then copies the contents into a new array (using toArray) as src/dst to do the actual copying? i.e. the source and destination seem disjoint.

Reply:
This comment was in the original patch - I'll fix or drop it.

Reply:
I think what you are seeing the extra copy is in the creation of the `truthSegment`, where an extra copy had to be used because there is no API (yet), to help the code determine which direction the copy should be performed in a self-copy scenario.  If the call to the MemoryCopy API does not perform the self-copy correctly, comparing its result with the `truthSegment` will definitely reveal an error.

Reply:
Actually, I'm talking about the call to toArray here (as well as for CopyTo): 

```
ByteOrder bo = mode.swap ? NON_NATIVE_ORDER : NATIVE_ORDER;
//CopyFrom
Object srcArr = helper.toArray(base); // <---------------------
int srcIndex = mode.direction ? 0 : indexShifts;
int srcCopyLen = helper.length(srcArr) - indexShifts;
```

This call to `toArray` ends of calling one of the `MemorySegment::to*Array` methods, which always copy the contents of the segment.

Reply:
You're right, I seem to have misread the code earlier. For some reason I thought a copy to/from the `srcArr` to/fromthe `base` segment was happening, but looking at it now, I can see it's doing as you describe. Sorry.